sql-plugin/src/main/scala/com/nvidia/spark/rapids/CostBasedOptimizer.scala

```diff
@@ -115,25 +118,50 @@ class CostBasedOptimizer(conf: RapidsConf) extends Logging {
     if (totalGpuCost > totalCpuCost) {
       // we have reached a point where we have transitioned onto GPU for part of this
       // plan but with no benefit from doing so, so we want to undo this and go back to CPU
-      if (plan.canThisBeReplaced) {
+      if (plan.canThisBeReplaced && !consumesQueryStage(plan)) {
         // this plan would have been on GPU so we move it and onto CPU and recurse down
         // until we reach a part of the plan that is already on CPU and then stop
         optimizations.append(ReplaceSection(plan, totalCpuCost, totalGpuCost))
-        plan.recursiveCostPreventsRunningOnGpu()
+        recursiveCostPreventsRunningOnGpu(plan)
       }
 
       // reset the costs because this section of the plan was not moved to GPU
       totalGpuCost = totalCpuCost
     }
 
-    if (!plan.canThisBeReplaced) {
+    if (!plan.canThisBeReplaced || consumesQueryStage(plan)) {
       // reset the costs because this section of the plan was not moved to GPU
       totalGpuCost = totalCpuCost
     }
 
     (totalCpuCost, totalGpuCost)
   }
 
+  /**
+   * Recursively force a section of the plan back onto CPU, stopping once a plan
+   * is reached that is already on CPU or cannot be moved back onto CPU.
+   */
+  final def recursiveCostPreventsRunningOnGpu(plan: SparkPlanMeta[_]): Unit = {
```

Comment:
This was supposed to be `private def` not `final def`. I will push a fix later today.